main.go

```diff
@@ -128,10 +341,161 @@ func determineShouldFetch(lastFetchedFilepath string) bool {
+	// Find TBD header
+    for scanner.Scan() {
+		if tbdHeaderRegex.Match(scanner.Bytes()) {
+			break
+		}
+    }
+	if err := scanner.Err(); err != nil {
+        return false, stacktrace.Propagate(err, "An error occurred while scanning for the TBD header in the changelog file at provided path: '%s'.\n", changelogFilepath)
+    }
+	// Scan file until next version header detected, searching for Breaking Changes header along the way
+	foundBreakingChanges := false
+	for scanner.Scan() {
+		if breakingChangesRegex.Match(scanner.Bytes()){
+			foundBreakingChanges = true
+		}
+		if versionHeaderRegex.Match(scanner.Bytes()){
+			break
+		}
+	}
+    if err := scanner.Err(); err != nil {
+		return false, stacktrace.Propagate(err, "An error occurred while scanning for the breaking changes header in the changelog file at provided path: '%s'.\n", changelogFilepath)
+    }
+
+    return foundBreakingChanges, nil
+}
+
+func runPreReleaseScripts(preReleaseScriptsDirpath string, releaseVersion string) error {
+	preReleaseScriptsFilepath := path.Join(preReleaseScriptsDirpath, preReleaseScriptsFilename)
+	preReleaseScriptsFile, err := os.ReadFile(preReleaseScriptsFilepath);
+	if err != nil {
+		return stacktrace.Propagate(err, "An error occurred attempting to open file at provided path. Are you sure '%s' exists?", preReleaseScriptsFilepath)
+	}
+
+	lines := bytes.Split(preReleaseScriptsFile, []byte("\n"))
+	var allScriptFilepaths []string
+	for _, line := range(lines) {
+		allScriptFilepaths = append(allScriptFilepaths, string(line))
+	}
+	for _, scriptFilepath := range allScriptFilepaths {
+		if strings.TrimSpace(scriptFilepath) == "" {
+			continue
+		}
+		scriptCmdString := path.Join(preReleaseScriptsDirpath, scriptFilepath)
+		scriptCmd := exec.Command(scriptCmdString, releaseVersion)
+		err := scriptCmd.Run()
+		if err != nil {
+			return stacktrace.Propagate(err, "An error occurred attempting to run the following pre release script command: '%s %s'", scriptCmdString, releaseVersion)
+		}
+	}
+
+	return nil
+}
+
+func updateChangelog(changelogFilepath string, releaseVersion string) error {
+	changelogFile, err := os.ReadFile(changelogFilepath)
+	if err != nil {
+		return stacktrace.Propagate(err, "An error attempting to open changelog file at provided path. Are you sure '%s' exists?", changelogFilepath)
+	}
+
+	lines := bytes.Split(changelogFile, []byte("\n"))
+	newLines:= make([][]byte, len(lines) + 2) 
+	// Add a new TBD header for next release
+	newLines[0] = []byte(tbdHeaderStr)
+	i := 1
+	for _, line := range lines {
+		// Change current TBD header to Release Version header
+		if tbdHeaderRegex.Match(line){
+			releaseVersionHeader := fmt.Sprintf("# %s", releaseVersion)
+			newLines[i + 1] = []byte(releaseVersionHeader)
+			i = i + 2
+		} else {
+			newLines[i] = line
+			i++
+		}
+	}
+	newChangelogFile := bytes.Join(newLines, []byte("\n"))
+	err = os.WriteFile(changelogFilepath, newChangelogFile, changelogFileMode)
```

Comment:
changelogFileMode can be read from the file, rather than hardcodd (which might change the file mode)